code/scripting/api/objs/ship.cpp

```diff
@@ -775,28 +775,22 @@ ADE_VIRTVAR(Target, l_Ship, "object", "Target of ship. Value may also be a deriv
 	else
 		return ade_set_error(L, "o", l_Object.Set(object_h()));
 
-	if(ADE_SETTING_VAR && newh)
-	{
-		if(aip->target_signature != newh->sig)
-		{
-			if(newh->IsValid())
-			{
-				aip->target_objnum = OBJ_INDEX(newh->objp);
-				aip->target_signature = newh->sig;
-				aip->target_time = 0.0f;
-
-				if (aip == Player_ai)
-					hud_shield_hit_reset(newh->objp);
-			}
-			else
-			{
-				aip->target_objnum = -1;
-				aip->target_signature = -1;
-				aip->target_time = 0.0f;
-			}
+	if(ADE_SETTING_VAR && !(newh && aip->target_signature == newh->sig)) {
+		// we have a different target, or are clearng the target
+		if(newh && newh->IsValid())	{
+			aip->target_objnum = OBJ_INDEX(newh->objp);
+			aip->target_signature = newh->sig;
+			aip->target_time = 0.0f;
 
-			set_targeted_subsys(aip, NULL, -1);
+			if (aip == Player_ai)
+				hud_shield_hit_reset(newh->objp);
+		} else if (lua_isnil(L, 2)) {
+			aip->target_objnum = -1;
+			aip->target_signature = -1;
+			aip->target_time = 0.0f;
 		}
```

Comment:
If neither of these two is the case (i.e., an invalid target object or not a target object) there should probably be a lua error. Or at least, it should return from the function instead of setting some targeted subsys to null.

Reply:
Ah, yes, `set_targeted_subsys` should be inside the blocks, but I didn't want to introduce errors for a situation where before it would've done nothing. I can if that's preferred, but that was my thinking.

Reply:
Okay, then probably ficing the set_targeted_subsys but not throwing any errors is probably best